### PR TITLE
Add ACR validation check for OIDC auth

### PR DESCRIPTION
**Pull request by crobby**
## Problem
This PR adds a check for an ACR value, which is an optional configuration.  It piggybacks on https://github.com/rancher/rancher/pull/45366
Also includes a fix to allow an empty Authentication URL, which is ok now since it will just trigger "discovery".

 
## Solution
If a required ACR value is present in the authconfig (supplied when configuring the auth provider), verify that the correct value is present in the token when the user logs in.
 
## Testing
Unit tests added

### Manual Testing
Configured Keycloak OIDC to provide an ACR value and then configured Rancher' authconfig to require it.  Logins with the correct ACR work while those without will fail.

### Automated Testing

* Test types added/modified:
    * Unit

### Regressions Considerations
This is added for the new genericoidc implementation, so I would not expect any existing functionality to be affected.

Existing / newly added automated tests that provide evidence there are no regressions:

### CODE REVIEW

File: pkg/auth/providers/genericoidc/genericoidc_provider.go [hunk 1]
```diff
@@ -131,12 +131,18 @@ func (g *GenOIDCProvider) groupToPrincipal(groupName string) v3.Principal {
...
 		authURL,
 		config["clientId"],
 		config["rancherUrl"],
 	)
+
+	if config["acrValue"] != nil {
+		redirectURL += fmt.Sprintf("&acr_values=%s", config["acrValue"])
+	}
+
+	return redirectURL
```
Comment (enrichman) [inline pkg/auth/providers/genericoidc/genericoidc_provider.go:145]: Not really part of the PR but are we ok about silently skipping/ignoring the error from FetchAuthURL? The redirect URL will be incomplete in that case.

File: pkg/auth/providers/oidc/oidc_provider.go [hunk 2]
```diff
@@ -402,6 +405,11 @@ func (o *OpenIDCProvider) getUserInfo(ctx *context.Context, config *v32.OIDCConf
 	if !reflect.DeepEqual(oauth2Token, reusedToken) {
 		o.UpdateToken(reusedToken, userName)
 	}
+	// If an ACR value is configured, validate the value in the claim
+	if config.AcrValue != "" && !validateACR(oauth2Token, config) {
```
Comment (enrichman) [inline pkg/auth/providers/oidc/oidc_provider.go:409]: We can probably skip the check for the emptiness, since it's done in the validate func.

File: pkg/auth/providers/oidc/oidc_provider.go [hunk 3]
```diff
@@ -515,3 +523,26 @@ func (o *OpenIDCProvider) getOIDCProvider(ctx context.Context, oidcConfig *v32.O
...
+	if !ok {
+		logrus.Errorf("acr value could not be parsed")
+		return false
+	}
+	if acr != config.AcrValue {
+		logrus.Infof("acr value in token does not match configured acr value")
+		return false
+	}
+	return true
+}
```
Comment (enrichman) [inline pkg/auth/providers/oidc/oidc_provider.go:548]: When returning a bool I usually prefer using the `IsXXX` convention, like `IsValidACR`. Not sure if in this case is still nice.

And it looks like we are not using anything else from the config, so we can probably pass directly the AcrValue:

```go
func IsValidACR(oauth2Token *oauth2.Token, acr string) bool {
    // ...
}
```

Btw looking at the code I think we can totally avoid this part.

**Review by enrichman — CHANGES_REQUESTED** [review]
The `ParseUnverified` was a bit odd, so having a look at the code I guess we can skip that part.

At line 390 we are already verifing the AccessToken, but dropping the result (maybe we didn't need it).

```go
idToken, err := verifier.Verify(updatedContext, oauth2Token.AccessToken)
```

We can now use it to get the `acr` claim though:

```go
var claims struct {
	ACR string `json:"oid"`
}

if err = idToken.Claims(&claims); err != nil {
	return "", fmt.Errorf("extracting claims: %w", err)
}

if config.AcrValue != "" {
	// check claims.ACR and config.AcrValue
}
```

What do you think?

I did something similar here:

https://github.com/enrichman/rancher/blob/995dd3d8d2c4a6a89e5a458a1789ce625f04cbc5/pkg/auth/providers/azure/clients/ms_graph_client.go#L344-L377

File: pkg/auth/providers/oidc/oidc_provider.go (hunk 2 above)
Comment (crobby) [reply pkg/auth/providers/oidc/oidc_provider.go:409]: done

File: pkg/auth/providers/genericoidc/genericoidc_provider.go (hunk 1 above)
Comment (crobby) [reply pkg/auth/providers/genericoidc/genericoidc_provider.go:145]: Since this is part of the shared code (the original keycloak oidc implementation also uses this), it's expected that we won't always have a configured ACR value.

File: pkg/auth/providers/oidc/oidc_provider.go (hunk 3 above)
Comment (crobby) [reply pkg/auth/providers/oidc/oidc_provider.go:548]: makes sense.  I've done some refactoring.

File: pkg/auth/providers/genericoidc/genericoidc_provider.go (hunk 1 above)
Comment (enrichman) [reply pkg/auth/providers/genericoidc/genericoidc_provider.go:145]: I was referring to the line before with `authURL, _ := baseoidc.FetchAuthURL(config)`, but because it was not part of the code changes, I was not able to select it.

I don't think it's a big issue, by the way. However, suppressing errors without returning or logging them catches my attention.

File: pkg/auth/providers/oidc/oidc_provider.go [hunk 4]
```diff
@@ -515,3 +531,36 @@ func (o *OpenIDCProvider) getOIDCProvider(ctx context.Context, oidcConfig *v32.O
...
+	}
+	if err := json.Unmarshal(decodedClaims, &tokenClaims); err != nil {
+		return "", fmt.Errorf("parsing token claims: %w", err)
+	}
+	acrValue, ok := tokenClaims["acr"].(string)
+	if !ok {
+		return "", errors.New("acr claim not found in token")
+	}
+	return acrValue, nil
+}
```
Comment (enrichman) [inline pkg/auth/providers/oidc/oidc_provider.go:566]: This manual decoding of the payload fails for some inputs, because of the `base64.URLEncoding`. You should use the `base64.RawURLEncoding` or  `base64.RawStdEncoding` instead, that are not expecting paddings in the encoded strings.

You can try with this random token:

```go
const accessToken = `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJhY3IiOiJmb29iYXIifQ.kKvbaR4j2N7FV1PWI1uH4YphC0KuL8YVsXDCzwcbSsg`
```

https://play.golang.com/p/z38g26yyYiB

If we are not going to use the returned verified token from the `_, err = verifier.Verify(updatedContext, oauth2Token.AccessToken)` at line 391 as suggested, and extracting manually the claims, then we can stick using the `ParseUnverified` of the jwt library. It's probably cleaner and safer:

```suggestion
func parseACRFromAccessToken(accessToken string) (string, error) {
	var parser jwt.Parser
	// we already validated the incoming token
	token, _, err := parser.ParseUnverified(accessToken, jwt.MapClaims{})
	if err != nil {
		return "", fmt.Errorf("failed to parse token: %w", err)
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return "", errors.New("invalid access token jwt.MapClaims format")
	}

	acrValue, found := claims["acr"].(string)
	if !found {
		return "", fmt.Errorf("acr claim invalid or not found in token: (acr=%v)", claims["acr"])
	}
	return acrValue, nil
}
```

https://play.golang.com/p/-2qFOF1IndC

**Review by enrichman — CHANGES_REQUESTED** [review]

**Review by enrichman — APPROVED** [review]

**Review by andreas-kupries — APPROVED** [review]

File: pkg/auth/providers/genericoidc/genericoidc_provider.go [hunk 5]
```diff
@@ -131,12 +131,18 @@ func (g *GenOIDCProvider) groupToPrincipal(groupName string) v3.Principal {
 func (g *GenOIDCProvider) getRedirectURL(config map[string]interface{}) string {
 	authURL, _ := baseoidc.FetchAuthURL(config)
 
-	return fmt.Sprintf(
+	redirectURL := fmt.Sprintf(
```
Comment (pmatseykanets) [inline pkg/auth/providers/genericoidc/genericoidc_provider.go:134]: Should we use [`url.JoinPath`](https://pkg.go.dev/net/url#JoinPath) here?

**Review by pmatseykanets — APPROVED** [review]

File: pkg/auth/providers/genericoidc/genericoidc_provider.go (hunk 5 above)
Comment (enrichman) [reply pkg/auth/providers/genericoidc/genericoidc_provider.go:134]: @pmatseykanets no, because those are query params. It could be possible to use `url.Values` and the [`url.Values.Encode`](https://pkg.go.dev/net/url#Values.Encode).

```go
values := url.Values{}
values.Set("client_id", "myclient")
values.Set("response_type", "code")
values.Set("redirect_uri", "https://google.com")

redirectURL := fmt.Sprintf("%s?%s", authURL, values.Encode())
```
